src/lib.rs

```diff
@@ -1021,57 +1019,121 @@ fn get_docker_for_tcp(tcp_host_str: String) -> Docker {
+    /// falling back to unix socket with path [DEFAULT_SOCKET_PATH](DEFAULT_SOCKET_PATH)) if
+    /// `unix-socket` feature is enabled.
+    ///
+    /// # Panics
+    ///
+    /// This function panics when a given url is incorrect or if the url specified is a unix socket
+    /// and `unix-socket` feature is disabled. To safely initialize Docker instance from
+    /// environment variable use [Docker::try_env](Docker::try_env] or
+    /// [Docker::try_new](Docker::try_new) that takes a url directly as parameter.
     pub fn new() -> Docker {
         match env::var("DOCKER_HOST").ok() {
-            Some(host) => {
-                #[cfg(feature = "unix-socket")]
-                if let Some(path) = host.strip_prefix("unix://") {
-                    return Docker::unix(path);
-                }
-                let host = host.parse().expect("invalid url");
-                Docker::host(host)
+            Some(host) => Docker::try_new(host).unwrap(),
+            #[cfg(feature = "unix-socket")]
+            None => Docker::unix(DEFAULT_SOCKET_PATH),
+            #[cfg(not(feature = "unix-socket"))]
+            None => {
+                panic!("unix socket not supported (enable feature `unix-socket` in Cargo.toml)")
             }
+        }
+    }
+    /// Constructs a new Docker instance based on uri.
+    ///
+    /// uri can be either a unix or tcp socket, for example `unix:///run/docker.sock` or for tcp
+    /// `tcp://127.0.0.1:80`.
+    ///
+    /// Valid schemes are: `tcp`, `http`, `https`, `unix`
+    pub fn try_new<S>(uri: S) -> Result<Docker>
+    where
+        S: Into<String>,
+    {
+        let uri_str = uri.into();
+
+        #[allow(clippy::single_match)] // this match is much nicer than two separate if lets
+        match uri_str.strip_prefix("unix://") {
             #[cfg(feature = "unix-socket")]
-            None => Docker::unix("/var/run/docker.sock"),
+            Some(path) => return Ok(Docker::unix(path)),
             #[cfg(not(feature = "unix-socket"))]
-            None => panic!("Unix socket support is disabled"),
+            Some(_) => return Err(Error::UnsupportedScheme { scheme: "unix" }),
+            _ => {}
+        }
+
+        let uri = uri_str.parse::<Uri>()?;
+
+        match uri.scheme().map(|scheme| scheme.as_str()) {
+            Some("tcp") | Some("http") | Some("https") => Docker::tcp(uri_str),
+            Some(s) => Err(Error::UnsupportedScheme { scheme: s.into() }),
+            None => Err(Error::UnsupportedScheme { scheme: "".into() }),
         }
     }
 
-    /// Creates a new docker instance for a docker host
-    /// listening on a given Unix socket.
+    /// Constructs a new Docker instance for a docker host listening at a url specified by an env var `DOCKER_HOST`,
+    /// falling back to unix socket with path [DEFAULT_SOCKET_PATH](DEFAULT_SOCKET_PATH)) if
+    /// `unix-socket` feature is enabled, otherwise returns an error.
+    pub fn try_env() -> Result<Docker> {
```

Comment:
I think `try_from_env()` is more obvious name. `try_env()` sounds like it's taking some action to me

Reply:
Agreed and done.